LSM: Snapshot sequence semantics simplification

sentientwaffle opened the pull request:
Modified for a more manageable mental model.

The progression of snapshots over successive compactions was difficult to visualize. From a messy mental model came bugs, contradictory documentation, and weak invariants.

### Snapshot semantics:
- (These are hopefully more thoroughly explained within the code+docs).
- `compaction.snapshot` is now the first beat of the compaction. (And renamed to `compaction.op_min`; more below).
  - This is neither the `snapshot_max` of the input tables, nor the `snapshot_min` of the output tables...
  - But it can easily derive both.
  - And it is even better, because it can't be confused for either. (I did this constantly before).
  - As a round number (a multiple of a half-bar start-beat) it is easier to visualize and remember (at least for me) than one-minus-what-measure-was-this-again. (Previously the compaction snapshot tied into the commit op of the data being compacted to disk, which is one measure behind the compact's op. But this is hard to reason about so I scrapped it. I document the (absence of) a relationship between snapshots and ops in `lsm/README.md`.)
- `lookup_snapshot_max` is now always the first beat of a half-bar, instead of sometimes being sometimes 1 less than that until it is incremented.
- Invariants now enforced:
  - In `lookup_from_memory()`, if the mutable table is visible, then the immutable table must also be visible. (This would have caught several earlier bugs).
  - A compaction's input tables must be visible to the `compaction.snapshot`.

### Bug fixes:
- `Tree.compact()`: When recovering from a checkpoint, we skip replaying one bar of compactions. But we still must compact the mutable → immutable table at the last beat of that (otherwise) skipped measure.
- `lsm/test.zig`: Don't checkpoint the same op twice.

### Naming:
- Rename `prefetch_snapshot_max` to `lookup_snapshot_max`. "Prefetching" is the Groove's concern — the Tree only knows about lookups. (Tree documentation still refers to prefetching because it is useful for understanding how lookups correlate to compaction ops).
- Rename "measure" to "bar" — same meaning, more concise.
- Rename `compaction.snapshot` to `compaction.op_min`. I was on the fence about this — `op_min` is a better description of where it comes from, but `snapshot` is a better description of how it is used. I settled on `op_min` because it is more specific ("snapshot" is very general). With `op_min` it is obvious what the value is.

### Documentation refactoring:
- Remove the over-complicated chart showing repeated compactions, in favor of a smaller, simpler, more general chart that is explored in more detail.
- Move the excellent "Tree compaction runs to the sound of music!" to a more prominent location.
  - Add a note about `compact` taking an op, since "compaction op" is referred to often.
  - Change "there's only levels/2 max compactions happening" to "there are at most levels/2 compactions running" (slightly more accurate).
  - Fix an out-of-date note about when/how the manifest info is updated.
  - Add a note about output table visibility.
- The `lookup_snapshot_max` chart is moved to be less prominent, since it is less critical for reasoning about snapshots. This chart is also simplified with a reduced emphasis on `compaction_op`, and benefits from the now-round `lookup_snapshot_max`.

sentientwaffle committed:
LSM: Snapshot sequence semantics simplification

Modified for a more manageable mental model.

The progression of snapshots over successive compactions was difficult to visualize. From a messy mental model came bugs, contradictory documentation, and weak invariants.

Snapshot semantics:
- (These are hopefully more thoroughly explained within the code+docs).
- `compaction.snapshot` is now the first beat of the compaction.
  - This is neither the `snapshot_max` of the input tables, nor the `snapshot_min` of the output tables...
  - But it can easily derive both.
  - And it is even better, because it can't be confused for either. (I did this constantly before).
  - As a round number (a multiple of a half-bar start-beat) it is easier to visualize and remember (at least for me) than one-minus-what-measure-was-this-again. (Previously the compaction snapshot tied into the commit op of the data being compacted to disk, which is one measure behind the compact's op. But this is hard to reason about so I scrapped it. I document the (absence of) a relationship between snapshots and ops at the top of "tree.zig").
- `lookup_snapshot_max` is now always the first beat of a half-bar, instead of sometimes being sometimes 1 less than that until it is incremented.
- Invariants now enforced:
  - In `lookup_from_memory()`, if the mutable table is visible, then the immutable table must also be visible. (This would have caught several earlier bugs).
  - A compaction's input tables must be visible to the `compaction.snapshot`.

Bug fixes:
- `Tree.compact()`: When recovering from a checkpoint, we skip replaying one bar of compactions. But we still must compact the mutable → immutable table at the last beat of that (otherwise) skipped measure.
- `lsm/test.zig`: Don't checkpoint the same op twice.

Naming:
- Rename `prefetch_snapshot_max` to `lookup_snapshot_max`. "Prefetching" is the Groove's concern — the Tree only knows about lookups. (Tree documentation still refers to prefetching because it is useful for understanding how lookups correlate to compaction ops).
- Begin to use "bar" instead of "measure" because it is shorter. (More of this to follow).

Documentation refactoring:
- Remove the over-complicated chart showing repeated compactions, in favor of a smaller, simpler, more general chart that is explored in more detail.
- Move the excellent "Tree compaction runs to the sound of music!" to a more prominent location.
  - Add a note about compact taking an op, since "compaction op" is referred to often.
  - Change "there's only levels/2 max compactions happening" to "there are at most levels/2 compactions running" (slightly more accurate).
  - Fix an out-of-date note about when/how the manifest info is updated.
  - Add a note about output table visibility.
- The `lookup_snapshot_max` chart is moved to be less prominent, since it is less critical for reasoning about snapshots. This chart is also simplified with a reduced emphasis on `compaction_op`, and benefits from the now-round `lookup_snapshot_max`.

sentientwaffle committed:
LSM: Refactor compaction/snapshot documentation

- `tree.zig`'s module comment was getting long; extract it to `src/lsm/README.md`.
- Rename "measure" to "bar" — same meaning, more concise.
- Rename `compaction.snapshot` to `compaction.op_min`. I was on the fence about this — `op_min` is a better description of where it comes from, but `snapshot` is a better description of how it is used. I settled on `op_min` because it is more specific ("snapshot" is very general). With `op_min` it is at least obvious what the value is.
- Elaborate the "Snapshot Values" documentation.
  - Previously this was hard to follow — it seemed to imply that `commit(X)` could not see the updates done by `commit(X-1)`.
  - Hopefully with the explained diagram this is easier to understand now.
- Remove the "beat" label from "(op, beat, snapshot)". Beats range from 0 to `lsm_batch_multiple-1` (and then wrap) — they are not the same as ops/snapshots.
- Explain why `lookup_snapshot_max` is the right value for the immutable table's `snapshot_min` at the end of the compaction bar.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+A compaction beat is started asynchronously with `compact()` which takes a callback and
+the op that was just committed.
+
+A bar is split in half according to the "first" down beat and "middle" down beat.
```
What's a down beat? What order do down/up beats happen in?

sentientwaffle replied:
I assume this is part of the music analogy :man_shrugging:. It is spelled out more explicity down below, lines 31-49.

jamii replied:
You could probably just remove `down` from this sentence. First beat and middle beat are pretty clear, just the 'down' that is confusing as not introduced yet.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+
+* At the end of every beat, there is space in mutable table for the next beat.
+* The manifest info for the compaction's input tables is updated at the end of the half-bar.
+* The manifest info for the compaction's output tables is updated during the compaction.
```
What is meant by 'during the compaction'? That seems like it could describe any point during this process, but it's described as an invariant which makes me think it means some specific point in time.

sentientwaffle replied:
Good point, this is ambiguous here. I'll elaborate.

sentientwaffle replied:
You're right, this isn't exactly an invariant. And I think the previous line isn't even correct :facepalm:. I'll remove both; these were supposed to be about

> * The compactions' output tables are not visible until the compaction has finished.

anyway, which _is_ correct.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+* At the end of every beat, there is space in mutable table for the next beat.
+* The manifest info for the compaction's input tables is updated at the end of the half-bar.
+* The manifest info for the compaction's output tables is updated during the compaction.
+* The manifest is compacted at the end of every beat.
```
Is this a different compaction to the compaction of the levels?

sentientwaffle replied:
Yes, the compaction of levels is about compacting the keys/values/tables. Compaction of the manifest is about compaction of the table metadata index, to clean up tables that have moved or been removed.

jamii replied:
Is the manifest compaction also incremental, or is it's size bounded?

sentientwaffle replied:
Incremental. Most beats it won't actually trigger.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+
+1. (first) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Allow level visible table counts to overflow if needed.
```
What is a 'level visible table count' and why would it overflow?

sentientwaffle replied:
Is this more clear?:

>     * Allow the per-level visible table limits to overflow if needed.

jamii replied:
What are the visible table limits?

sentientwaffle replied:
I'll add these (in different sections):
```
- Level `0` … level `config.lsm_levels - 1` each contain an exponentially increasing number of
  on-disk tables: `config.lsm_growth_factor ^ (level + 1)`. (`config.lsm_growth_factor` is usually 8).
```

and:

```
    * Allow the per-level table limits to overflow if needed (for example, if we may compact a table
      from level `A` to level `B`, where level `B` is already full).
```

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+To avoid write amplification stalls and bound latency, compaction is done incrementally.
+
+A full compaction phase is denoted as a bar or measure, using terms from music notation.
+Each bar consists of `lsm_batch_multiple` beats or "compaction ticks" of work.
```
When do these beats happen? Who drives them?

sentientwaffle replied:
Is that not explained sufficiently by the next sentence? Or do you mean at a higher level, what is driving `compact()`?

jamii replied:
Yeah. I get the sense that 1 committed op = 1 beat, but I don't think that's explicitly spelled out here. My understanding is that every `lsm_batch_multiple` ops a new compaction is started and runs concurrently with op commits. If true, that would be good context to start with.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+    * Start even level compactions if there's any tables to compact.
+
+2. (second) up beat of the bar:
+    * Finish ticking running even-level compactions.
```
'ticking running'

sentientwaffle replied:
Ah, "running" is being used as an adjective of "compactions". I'll reword this, it is confusing.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+
+3. (third) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Start odd level compactions if there are any tables to compact, and only if we must.
```
Why might we must? And why not always do odd-level compactions?

sentientwaffle replied:
> Why might we must?

I'm not sure what this is referring to, actually. I'll clean this up.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+3. (third) down beat of the bar:
+    * Assert no compactions are currently running.
+    * Start odd level compactions if there are any tables to compact, and only if we must.
+    * Compact the immutable table if it contains any sorted values (it could be empty).
```
What are the mutable and immutable tables?

sentientwaffle replied:
Adding:

```
- The [mutable table](table_mutable.zig) is an in-memory table.
  - All tree updates, inserts, and removes are applied to the mutable table.
  - The mutable table's size is allocated to accommodate a full bar of updates.
- The [immutable table](table_immutable.zig) is an in-memory table.
  - The mutable table's contents are periodically moved to the immutable table,
    where they are stored while being flushed to level `0`.
```

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+    * Compact the immutable table if it contains any sorted values (it could be empty).
+
+4. (fourth) last beat of the bar:
+    * Finish ticking running odd-level and immutable table compactions.
```
'ticking running'

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+Tree compaction runs to the sound of music!
+
+Compacting LSM trees involves merging and moving tables into the next levels as needed.
+To avoid write amplification stalls and bound latency, compaction is done incrementally.
```
It's not obvious to me how this is incremental. Presumably we do some bounded amount of work per committed op, but from reading the description it's not ruled out that we might get a single op and then have to compact every level of the tree. Is there any bound on how many levels get compacted per op? Or are the compactions partial, and if so how do we guarantee that they always finish in time?

sentientwaffle replied:
We might compact every level on a single op, yes. The incremental nature is that each compaction (from level A to level A+1) is split up to take place across `lsm_batch_multiple` ops.

> Or are the compactions partial, and if so how do we guarantee that they always finish in time?

Yep. Right now we guarantee this at the last beat of each compaction, which will not return control until it finishes. I'll leave this documentation as-is for now — I'm currently working on fixing compaction scheduling to spread work more evenly between beats.

jamii replied:
So the maximum compaction stall is when the largest-but-one level compacts into the largest level, divided by `lsm_batch_multiple` ops? That would be ~1/2 the database in reads+writes over 4 ops, if I'm understanding this correctly.

sentientwaffle replied:
- Each table has the same maximum size, regardless of whether it is a level-0 table or a level-N table.
- The number of tables increases exponentially at each successive level.
- So the key range spanned by a table in level `X+1` is typically smaller than the key range spanned by a table in level `X`.

So what that is the worst case stall, but it would require a weird write workload to trigger — on average 1/8 of the level-B tables will overlap with the level-A table. With how our actual trees are indexed (using a timestamp or a UUID) it is probably impossible, but I haven't worked through it entirely. (Unless you chose UUIDs antagonistically :stuck_out_tongue:).

(Also, although we currently set `lsm_batch_multiple` to 4, it will be set higher in production.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+    * Flush, clear, and sort mutable table values into immutable table for next bar.
+
+
+## Snapshots
```
What is a snapshot and why do we want one?

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+## Snapshots
+### Snapshots and Compaction
+
+Consider the half-bar compaction beginning at op=`X` (`12`), with `lsm_batch_multiple=M` (`8`).
```
The explanation above only covered M=4. How does the work get split up for M=8?

sentientwaffle replied:
I will rework the compaction explanation; that explanation doesn't rely on the batch multiple at all.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+
+During the compaction `X` (op=`X…Y-1`; `12…15`), each commit prefetches from the snapshot equal to
+the first op of the compaction. As shown, they continue to query the old (input) tables.
+
```
...and this is ok (no stale reads) because the mutable table doesn't get flushed until the end of the compact?

sentientwaffle replied:
Conceptually the mutable table is an "input table", to the immutable table's "output table". (But of course their compaction is just a sort+copy). This relationship is explored more in the "Snapshot Values" section.

jamii replied:
I guess what's missing is knowing that ops read and write to the mutable table, and that the number of writes per op is bounded so we know we can go a whole half-bar without flushing the mutable table.

sentientwaffle replied:
Correct, except that we go a full bar without flushing the mutable table, not a half-bar. I'll detail this in the docs.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+- The output tables of the above compaction `X` are visible.
+- The input tables of the above compaction `X` are invisible.
+- Therefore, it will lookup from the output tables, but ignore the input tables.
+- Commits must not prefetch from the output tables of `X` before the compaction half-bar
```
'before the compaction half-bar has finished' = before Y?

sentientwaffle replied:
Not quite, but it's subtle. I'll clarify that to:
```
- Callers must not query from the output tables of `X` before the compaction half-bar has finished
  (i.e. before the end of beat `Y-1` (`15`)), since those tables are incomplete.
```

I also generalize from prefetching to more general queries, since from an abstraction-boundary point-of-view the `Tree` doesn't know anything about prefetches, just lookups.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+- Commits must not prefetch from the output tables of `X` before the compaction half-bar
+  has finished, since those tables are incomplete.
+
+At this point the input tables can be removed if they are invisible to all persistent snapshots.
```
What is a persistent snapshot? I understand the input/output distinction for allowing ops to read/write while the compaction is in process, but why would we hang onto the input table after the compaction finished and the mutable table flushed?

sentientwaffle replied:
Persistent snapshot's aren't implemented yet, though much of the machinery is there. The idea is similar to how snapshots are implemented in some file systems — you can query the past state of the database (i.e. time travel). e.g. "what was the balance of account X at time Y". (Though you won't actually be able to do this for arbitrary times, only for times with persisted snapshots).

i.e. we wouldn't hang on to _all_ input tables, only ones that are visible to the handful of persisted snapshots.

jamii commented on src/lsm/README.md:
````diff
@@ -0,0 +1,111 @@
+```
+
+Compaction is driven by the commits of ops `B→C` (`16…23`):
+- Updates from ops `A→B` (`8…15`) are in the immutable table.
````
At what point in time is this statement true? In snapshot B?

sentientwaffle replied:
All the way from the beginning of op 16 (or really, the end of op 15) to 23 (they are reset at the end of 23). I'll expand on this.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+Compaction is driven by the commits of ops `B→C` (`16…23`):
+- Updates from ops `A→B` (`8…15`) are in the immutable table.
+- Updates from ops `B→C` (`16…23`) are added to the mutable table (by the respective commit).
+- `lookup_snapshot_max` is `B` (`16`).
```
What is this variable?

sentientwaffle replied:
Rewritten to:
```
- Lookups target snapshot `B` (`16`) (`Tree.lookup_snapshot_max`).
```

jamii replied:
Is this the snapshot against which op commits will do reads?

sentientwaffle replied:
Yes, commits always read against `tree.lookup_snapshot_max` (ignoring persistent snapshots).

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+- Updates from ops `A→B` (`8…15`) are in the immutable table.
+- Updates from ops `B→C` (`16…23`) are added to the mutable table (by the respective commit).
+- `lookup_snapshot_max` is `B` (`16`).
+  (Lookups against this snapshot read from the mutable and immutable tables (`A→C`; `8…23`) —
```
'this snapshot' = snapshot C?

sentientwaffle replied:
`B` actually. I'll elaborate on `lookup_snapshot_max`.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+During the compaction `X`:
+- `snapshot_max` of each input table is truncated to `Y-1` (`15`).
+- `snapshot_min` of each output table is initialized to `Y` (`16`).
+
```
What are `snapshot_min`/`snapshot_max`?

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+Assuming a bar/`lsm_batch_multiple` of 4, the invariants can be described as follows:
+
+* At the end of every beat, there is space in mutable table for the next beat.
+* The manifest info for the compaction's input tables is updated at the end of the half-bar.
```
How do tables and levels relate? Does table contain a list of pointers to levels on disk?

sentientwaffle replied:
I will need to write something up on what a manifest is, but that is probably out of scope for this PR. Each table lives in a single level. The manifest is an index that tracks the location (i.e. the level) of every table, along with other table metadata.

jamii replied:
Also out of scope, but could you put a vocab section at the top of the readme? Could maybe include some of lsm, forest, groove, tree, table (mutable/immutable), level, manifest, beat, bar, measure (same as a bar?), snapshot.

Or I could write it in a separate PR, but I'll have to ping you for some of them.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,111 @@
+
+### Snapshot Values
+
+The on-disk tables visible to a snapshot `B` do not contain the updates from the commit with op `B`.
```
Why not?

sentientwaffle replied:
Because we tried that before and it was very hard to reason about :wink:. Surprisingly (for me anyway) "which ops' updates are in this snapshot" isn't useful at this level of LSM-tree abstraction.

jamii replied:
So the mutable table is flushed before op `B` to produce snapshot `B`?

sentientwaffle replied:
No, the mutable table flushed before `B` would produce snapshot `C`. But now that I write that out, I realize it is confusing. I'm going to see whether it would make sense to shift things by a bar so they match, or if it would be confusing in a different way. :thinking:

jamii reviewed (commented):
My notes from reading:

```
lsm
  mutable table (in-memory)
  immutable table (in-memory, sorted)
  levels (on-disk, sorted)

we have to compact incrementally to avoid stalls
  for each committed op, we have a bound on how many values it can write
  for N values written, there is a bound on how much compaction work can be required
  so we do enough work per `lsm_batch_multiple` committed ops that we never fall behind

we want to compact concurrently to committing ops
  => ops need to read from tables that are currently being compacted

a new compaction starts every lsm_batch_multiple ops
  turns old table into new table - same logical contents but compact-er layout on disk
op's commit logic writes to the mutable table
op's commit logic reads from the old table and the mutable/immutable tables
during the compaction, we flush the immutable table to disk
at the end of the compaction, we replace the old table with the new table and flush the mutable table into the immutable table
every op in between gates some async work
  TODO not clear how much work is gated per op when lsm_batch_multiple != 4

TODO what exactly is a compaction?
do we compact an entire level at once or is it somehow incremental?
if all at once, will we get stalls from compacting the bigger levels?
if incremental, how do we guarantee that we do enough work to keep up?
```

sentientwaffle commented:
```

TODO what exactly is a compaction?
do we compact an entire level at once or is it somehow incremental?
if all at once, will we get stalls from compacting the bigger levels?
if incremental, how do we guarantee that we do enough work to keep up?
```

I will expand on this a bit in this PR, but more in a later one, since I am reworking compaction pacing anyway.

But short answer is, compaction is incremental — we select a table from level A, and compact it into the next level.

sentientwaffle committed:
LSM: Docs: Structure, compaction, snapshots

jamii commented on src/lsm/README.md:
```diff
@@ -1,4 +1,35 @@
+  - Each tree has as many as `config.lsm_growth_factor ^ (level + 1)` tables per level.
+    (`config.lsm_growth_factor` is typically 8).
+  - Within a given level and snapshot, the tables' key ranges are [disjoint](manifest_level.zig).
+
```
This is perfect, thanks :)

jamii commented on src/lsm/README.md:
````diff
@@ -8,104 +39,143 @@ To avoid write amplification stalls and bound latency, compaction is done increm
+        ↑A      ↑B      ↑C
 ```
 
 Compaction is driven by the commits of ops `B→C` (`16…23`):
````
```suggestion
Compaction is driven by the commits of ops `B→C` (`16…23`). While these ops are being committed:
```

sentientwaffle committed:
LSM: Snapshots: Prefetch from snapshot=op

Under normal operation (i.e. excluding the full bar following recovery from checkpoint) the commit of op `X` now prefetches from snapshot `X`. This doesn't require a change to the compaction table input/output snapshot min/max.

jamii commented on src/lsm/README.md:
```diff
@@ -0,0 +1,215 @@
+3. Second half-bar, first beat ("middle beat"):
+    * Assert no compactions are currently running.
+    * Start odd level compactions if there are any tables to compact.
+    * Start compactions from odd levels that have reached their table limit.
```
Are these two points distinct?

sentientwaffle replied:
No, it looks like I reworded it but left the old version in :facepalm:.

sentientwaffle replied:
The second is my (attempted) improvement, to be more explicit about which is the source level, and also that "any tables to compact" refers to.

jamii reviewed (dismissed):
This is so good.

I left one more nitpicky comment, but otherwise everything looks great.

sentientwaffle commented:
I amended the last commit the remove the redundant line you found. Can you re-approve the PR so it is mergable? And thanks for the great review!
